### PR TITLE
Cmake files and installation docs for Arpack

### PR DESCRIPTION
**Pull request by GitHimanshuc**
## Proposed changes

This PR contains shell scripts to install Arpack on clusters and cmake files to find and link it. 

#### For container users
 #3323 has been merged so the new containers already have Arpack installed, the end user only needs to pull the latest image.

#### For cluster users
Run spectre_setup_modules. All the shell files have been updated to install Arpack automatically and everything should just work. (Installation shell scripts have been updated for expanse, wheeler and CaltechHPC in this PR. Users of other clusters please make the relevant changes and open a PR for your cluster.)

#### For unix users
Install Arpack using your package manager.
                           **OR**
Build Arpack from source following the installation instructions given at https://github.com/opencollab/arpack-ng, and set the env variable ARPACK_ROOT to the installation directory.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by GitHimanshuc** [comment]
## Questions:

- [ ] Is the location of the Arpack source files correct?
- [ ] There were a lot of things in the source code that were not inline with the formatting rules, I skipped those checks for now. What is the policy for dealing with formatting of the external source files?

## Notes:
- [ ] For the test Unit.ApparentHorizons.StrahlkorperGr.DimensionfulSpinMagnitude tolerance was reduced to $10^{-12}$ for it to pass.

## TODO
- [ ] Add licence files

**Comment by kidder** [comment]
Why are we copying this package and not using arpack-ng https://github.com/opencollab/arpack-ng which is available through spack?

**Comment by wthrowe** [comment]
Following up on a comment I made in today's meeting: I think the dependency info in my distro is wrong and scipy actually uses a bundled copy of arpack, not an installed one.  This is much less likely to cause conflicts with anything we do.

That said, I would still prefer using a version installed from upstream.

File: cmake/FindArpack.cmake
Comment (nilsvu) [inline cmake/FindArpack.cmake:10]: Could you split the changes that find the Arpack library and make it available in CMake into a separate commit please? It'll also need an entry in `docs/Installation/Installation.md`.

Question: Are you suggesting to make Arpack an optional dependency, or will it be required? It's probably easy enough to install to make it a required dependency, but we should make sure of that.

File: cmake/FindArpack.cmake
Comment (nilsvu) [inline cmake/FindArpack.cmake]: This should be `ARPACK_ROOT`

File: cmake/SetupArpack.cmake
Comment (nilsvu) [inline cmake/SetupArpack.cmake]: not needed, this is handled by `find_package`. Only keep line 9.

File: cmake/SetupArpack.cmake
Comment (nilsvu) [inline cmake/SetupArpack.cmake]: You aren't getting these linker flags from anywhere, so the variable is empty and this command doesn't do anything. Are they needed?

File: cmake/SetupArpack.cmake
Comment (nilsvu) [inline cmake/SetupArpack.cmake:17]: Also find the include directory in `FindArpack` and add it to the target here.

**Review by nilsvu — CHANGES_REQUESTED** [review]
I only reviewed the cmake files.

File: docs/Installation/Installation.md
Comment (nilsvu) [inline docs/Installation/Installation.md]: Maybe move this up, below "LAPACK"?

Comment (nilsvu) [reply docs/Installation/Installation.md]: Arpack will need to be installed on clusters. @kidder @wthrowe Do you have suggestions for doing that?

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: docs/Installation/Installation.md
Comment (nilsvu) [reply docs/Installation/Installation.md]: And third: Please add a few notes on the new required dependency and some brief instructions to the "upgrade instructions" section of the PR description. This will help people upgrade their build environments once this PR is merged.

File: support/Environments/wheeler_gcc.sh
Comment (GitHimanshuc) [inline support/Environments/wheeler_gcc.sh]:
```suggestion
    module load arpack
```

**Comment by GitHimanshuc** [comment]
I have checked this PR on wheeler, Caltech HPC and expanse.

**Comment by GitHimanshuc** [comment]
### CMakeError.log

`Determining if the Fortran sgemm exists failed with the following output:
Change Dir: /central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake cmTC_27c06/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_27c06.dir/build.make CMakeFiles/cmTC_27c06.dir/build
gmake[1]: Entering directory `/central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp'
Building Fortran object CMakeFiles/cmTC_27c06.dir/testFortranCompiler.f.o
/software/gcc/9.2.0/bin/gfortran    -c /central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_27c06.dir/testFortranCompiler.f.o
Linking Fortran executable cmTC_27c06
/central/software/cmake/3.18.0/bin/cmake -E cmake_link_script CMakeFiles/cmTC_27c06.dir/link.txt --verbose=1
/software/gcc/9.2.0/bin/gfortran CMakeFiles/cmTC_27c06.dir/testFortranCompiler.f.o -o cmTC_27c06 
CMakeFiles/cmTC_27c06.dir/testFortranCompiler.f.o: In function `MAIN__':
testFortranCompiler.f:(.text+0xa): undefined reference to `sgemm_'
collect2: error: ld returned 1 exit status
gmake[1]: *** [cmTC_27c06] Error 1
gmake[1]: Leaving directory `/central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp'
gmake: *** [cmTC_27c06/fast] Error 2



Performing C SOURCE FILE Test CMAKE_HAVE_LIBC_PTHREAD failed with the following output:
Change Dir: /central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake cmTC_bea1e/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_bea1e.dir/build.make CMakeFiles/cmTC_bea1e.dir/build
gmake[1]: Entering directory `/central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_bea1e.dir/src.c.o
/software/gcc/9.2.0/bin/gcc   -DCMAKE_HAVE_LIBC_PTHREAD -o CMakeFiles/cmTC_bea1e.dir/src.c.o -c /central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp/src.c
Linking C executable cmTC_bea1e
/central/software/cmake/3.18.0/bin/cmake -E cmake_link_script CMakeFiles/cmTC_bea1e.dir/link.txt --verbose=1
/software/gcc/9.2.0/bin/gcc  -DCMAKE_HAVE_LIBC_PTHREAD CMakeFiles/cmTC_bea1e.dir/src.c.o -o cmTC_bea1e 
CMakeFiles/cmTC_bea1e.dir/src.c.o: In function `main':
src.c:(.text+0x2d): undefined reference to `pthread_create'
src.c:(.text+0x39): undefined reference to `pthread_detach'
src.c:(.text+0x45): undefined reference to `pthread_cancel'
src.c:(.text+0x56): undefined reference to `pthread_join'
src.c:(.text+0x6a): undefined reference to `pthread_atfork'
collect2: error: ld returned 1 exit status
gmake[1]: *** [cmTC_bea1e] Error 1
gmake[1]: Leaving directory `/central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp'
gmake: *** [cmTC_bea1e/fast] Error 2


Source file was:
#include <pthread.h>

void* test_func(void* data)
{
  return data;
}

int main(void)
{
  pthread_t thread;
  pthread_create(&thread, NULL, test_func, NULL);
  pthread_detach(thread);
  pthread_cancel(thread);
  pthread_join(thread, NULL);
  pthread_atfork(NULL, NULL, NULL);
  pthread_exit(NULL);

  return 0;
}

Determining if the function pthread_create exists in the pthreads failed with the following output:
Change Dir: /central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake cmTC_2ae5b/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_2ae5b.dir/build.make CMakeFiles/cmTC_2ae5b.dir/build
gmake[1]: Entering directory `/central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_2ae5b.dir/CheckFunctionExists.c.o
/software/gcc/9.2.0/bin/gcc   -DCHECK_FUNCTION_EXISTS=pthread_create -o CMakeFiles/cmTC_2ae5b.dir/CheckFunctionExists.c.o -c /central/software/cmake/3.18.0/share/cmake-3.18/Modules/CheckFunctionExists.c
Linking C executable cmTC_2ae5b
/central/software/cmake/3.18.0/bin/cmake -E cmake_link_script CMakeFiles/cmTC_2ae5b.dir/link.txt --verbose=1
/software/gcc/9.2.0/bin/gcc  -DCHECK_FUNCTION_EXISTS=pthread_create CMakeFiles/cmTC_2ae5b.dir/CheckFunctionExists.c.o -o cmTC_2ae5b  -lpthreads 
/usr/bin/ld: cannot find -lpthreads
collect2: error: ld returned 1 exit status
gmake[1]: *** [cmTC_2ae5b] Error 1
gmake[1]: Leaving directory `/central/groups/sxs/hchaudha/spectre/build_debug/CMakeFiles/CMakeTmp'
gmake: *** [cmTC_2ae5b/fast] Error 2
`

**Comment by nilsvu** [comment]
The install scripts look good, though please add them only for the ones you have tested. For example, it won't work on Minerva, but you can leave updating Minerva to me :) Similarly, @geoffrey4444 will surely handle installing arpack on Ocean.

File: cmake/FindArpack.cmake
Comment (nilsvu) [inline cmake/FindArpack.cmake]: I think checking for `arpack/arpack.hpp` here is the right thing to do, which is the file you would include for the C++ bindings, no?

File: cmake/FindArpack.cmake
Comment (nilsvu) [inline cmake/FindArpack.cmake]: remove `include` here

File: cmake/SetupArpack.cmake
Comment (nilsvu) [inline cmake/SetupArpack.cmake:9]: Please add this as well:

```cmake
file(APPEND
  "${CMAKE_BINARY_DIR}/BuildInfo.txt"
  "ARPACK_LIBRARIES: ${ARPACK_LIBRARIES}\n"
  "ARPACK_INCLUDE_DIR: ${ARPACK_INCLUDE_DIR}\n"
  )
```

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: cmake/FindArpack.cmake
Comment (GitHimanshuc) [reply cmake/FindArpack.cmake]: Done.
We are not using the c++ bindings provided by arpack-ng because there are quite recent bug fixes in the git repo. We are calling Arpack directly `src/NumericalAlgorithms/LinearAlgebra/FindGeneralizedEigenvaluesArpack.cpp` #3431.

File: cmake/FindArpack.cmake
Comment (GitHimanshuc) [reply cmake/FindArpack.cmake]: Done

File: cmake/SetupArpack.cmake
Comment (GitHimanshuc) [reply cmake/SetupArpack.cmake:9]: Added

**Review by nilsvu — DISMISSED** [review]
LGTM 👍

@kidder @wthrowe @markscheel Could one of you please test this on Caltech HPC and Wheeler? Does one of you volunteer to add arpack to the Frontera and Expanse env files?

@geoffrey4444 Can you handle Ocean? I'll request your review here, so we merge this only once you have given your OK.

File: cmake/FindArpack.cmake
Comment (wthrowe) [reply cmake/FindArpack.cmake]: If you're not using the C++ bindings, why check for them?

**Review by wthrowe — CHANGES_REQUESTED** [review]
See above thread on C++ bindings.

File: cmake/FindArpack.cmake
Comment (nilsvu) [reply cmake/FindArpack.cmake]: I think it's better to find the include dir, then we have the possibility to use the includes. Which header we look for is probably not super important. My suggestion is to leave as-is.

I don't see any recent commits, issues or PRs in the arpack-ng repo mentioning bugs, which are you referring to?

Comment (nilsvu) [reply cmake/FindArpack.cmake]: I reverse my position. Reading https://github.com/opencollab/arpack-ng/blob/master/CMakeLists.txt it looks like C++ bindings are not built by default. I agree then, that it's probably easiest to not rely on them and you can delete the include-dir find in cmake after all. I find this quite unfortunate though :(

Comment (wthrowe) [reply cmake/FindArpack.cmake]: My distro packages arpack-ng, but doesn't install the C++ bindings.  If we're not using the C++ bindings, I'd rather not be forced to re-package the thing just to satisfy a CMake check.

Comment (wthrowe) [reply cmake/FindArpack.cmake]: (Collision on the last two comments.)  I will say that I don't object to testing for the C bindings, since they should presumably be everywhere.  I also don't object to not testing for any headers, since we don't use them.

Comment (nilsvu) [reply cmake/FindArpack.cmake]: Agreed, see above. Thanks for checking this @wthrowe.

Comment (nilsvu) [reply cmake/FindArpack.cmake]: It looks like by default arpack installs no useful headers, so let's not try to find any. Even `spack install arpack-ng` doesn't install any.

Comment (GitHimanshuc) [reply cmake/FindArpack.cmake]: I will go ahead and remove the include-dir part from the cmake file.


@nilsleiffischer I was referring to bug fix commits in the ICB part of the code https://github.com/opencollab/arpack-ng/tree/master/ICB. There are quite a few in the last two years. The official ubuntu20.04 package for arpack(https://ubuntu.pkgs.org/20.04/ubuntu-universe-arm64/libarpack2-dev_3.7.0-3_arm64.deb.html) is 3years old which means a lot of these fixes may not be present.
Note that these binding are something that arpack-ng people have added in the last few years, the Arpack FORTRAN code that we calling directly is quite old and well tested.

Comment (nilsvu) [reply cmake/FindArpack.cmake]: Thanks for explaining 👍

**Comment by nilsvu** [comment]
LGTM, feel free to squash

**Review by nilsvu — DISMISSED** [review]

File: support/Environments/wheeler_clang.sh
Comment (markscheel) [inline support/Environments/wheeler_clang.sh]: This might be for another PR, but we probably want to change the message to say "..., so if you want this SpECTRE environment to be your primary environment, you may want to add it to your ~/.bashrc."  Same wording as in the file for expanse.  Or maybe even "primary" to "only".

(This is because if you do add it to your ./bashrc, it wreaks havoc if you want to compile SPECTRE in one environment while also running SPECTRE jobs in a different environment).

File: support/Environments/wheeler_gcc.sh
Comment (markscheel) [inline support/Environments/wheeler_gcc.sh]: Similar comment about adding to your .bashrc.

**Review by wthrowe — DISMISSED** [review]

File: support/Environments/wheeler_clang.sh
Comment (nilsvu) [reply support/Environments/wheeler_clang.sh]: Sounds reasonable, but let's do this in another PR 👍

**Comment by nilsvu** [comment]
@geoffrey4444 could you confirm that you can install arpack on Ocean? Just making sure here that this will not disrupt your workflows.

File: cmake/FindArpack.cmake
Comment (nilsdeppe) [reply cmake/FindArpack.cmake:10]: Related: the commits should be verby (bottom of https://spectre-code.org/contributing_to_spectre.html) eg "Add CMake find files for arpack", "Add Arpack to HPC caltech env"

File: support/Environments/wheeler_gcc.sh
Comment (nilsdeppe) [inline support/Environments/wheeler_gcc.sh]: you can delete everything you've added here. I've added an ARPACK module on wheeler. same in the clang config. Please confirm it works :)

File: docs/Installation/InstallationOnClusters.md
Comment (nilsdeppe) [inline docs/Installation/InstallationOnClusters.md]: This is not true anymore

Edit: as in, you can drop the changes you've made

File: docs/Installation/InstallationOnClusters.md
Comment (nilsdeppe) [inline docs/Installation/InstallationOnClusters.md:88]: @geoffrey4444 do you just want to install arpack as a system module on ocean?

File: support/Environments/wheeler_gcc.sh
Comment (GitHimanshuc) [reply support/Environments/wheeler_gcc.sh]: Thanks for installing Arpack on wheeler. I have removed the part that installed Arpack on wheeler and checked that the new version works.

**Comment by knelli2** [comment]
@GitHimanshuc  Since it's been a while, if you'd like to have these changes merged into spectre, please comment here or let the core-devs know. If you would no longer like to have these changes merged, or we don't hear from you by next Wed Aug 28 1pm PDT, we will close this PR. If you change your mind in the future, we can always re-open the PR.

**Pull request closed without merging**
